// @jsmini/is/dist/index.js
/*!
 * @jsmini/is 0.8.5 (https://github.com/jsmini/is)
 * API https://github.com/jsmini/is/blob/master/doc/api.md
 * Copyright 2017-2019 jsmini. All Rights Reserved
 * Licensed under MIT (https://github.com/jsmini/is/blob/master/LICENSE)
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var type = require('@jsmini/type');

function isInRange(x, min, max) {
  x = +x;
  min = +min;
  max = +max; // x 不是数字，则返回false

  if (isNaN(x)) return false; // min 或 max 不传，则认为不设置下限或上限

  return (!isNaN(min) ? x >= min : true) && (!isNaN(max) ? x <= max : true);
}
function isNumber(x, min, max) {
  return type.type(x) === 'number' && isInRange(x, min, max);
}
function isInteger(x, min, max) {
  return parseInt(x, 10) === x && isInRange(x, min, max);
}
function isInt(x) {
  // -2^31 ~ 2^31-1
  return isInteger(x, -2147483648, 2147483647);
}
function isBoolean(x) {
  return type.type(x) === 'boolean';
}
function isString(x) {
  return type.type(x) === 'string';
}
function isEmptyString(x) {
  if (!isString(x)) return false;
  return /^\s*$/.test(x);
}
function isNull(x) {
  return type.type(x) === 'null';
}
function isUndefined(x) {
  return type.type(x) === 'undefined';
}
function isObject(x) {
  return type.type(x) === 'object';
}
function isFunction(x) {
  return type.type(x) === 'function';
}
var isArray = isFunction(Array.isArray) ? Array.isArray : function isArray(x) {
  return type.type(x) === 'array';
};

exports.isInRange = isInRange;
exports.isNumber = isNumber;
exports.isInteger = isInteger;
exports.isInt = isInt;
exports.isBoolean = isBoolean;
exports.isString = isString;
exports.isEmptyString = isEmptyString;
exports.isNull = isNull;
exports.isUndefined = isUndefined;
exports.isObject = isObject;
exports.isFunction = isFunction;
exports.isArray = isArray;
